andi/andi.py

```diff
@@ -50,13 +68,271 @@ def to_provide(
+    >>> assert type(c) is C
+    >>> assert c.a is instances[A]
+    >>> assert c.b is instances[B]
+    >>> assert c.b.a is instances[A]
+    """
+    assert isinstance(cls, type)
+    is_injectable, externally_provided = _ensure_input_type_checks_as_func(
+        is_injectable, externally_provided)
+    return _plan(cls, is_injectable, externally_provided, True, None)
+
+
+def _plan(class_or_func: Union[Type, Callable],
+          is_injectable: Callable[[Type], bool],
+          externally_provided: Callable[[Type], bool],
+          strict,
+          dependency_stack=None) -> Plan:
+    dependency_stack = dependency_stack or []
+    plan_seq = OrderedDict()  # type: Plan
+    type_for_arg = {}
+
+    input_is_type = isinstance(class_or_func, type)
+
+    if input_is_type:
+        cls = cast(Type, class_or_func)
+        if externally_provided(cls):
+            plan_seq[cls] = {}
+            return plan_seq
+
+        if not is_injectable(cls):
+            raise NonProvidableError(
+                "Type {} cannot be provided".format(as_class_names(cls)))
+
+
+        if cls in dependency_stack:
+            raise CyclicDependencyError(
+                "Cyclic dependency found. Dependency graph: {}".format(
+                    " -> ".join(as_class_names(dependency_stack + [cls]))))
+        dependency_stack = dependency_stack + [cls]
+        arguments = inspect(cls.__init__)
+    else:
+        arguments = inspect(class_or_func)
+
+    for argname, types in arguments.items():
+        sel_cls = _select_type(types, is_injectable, externally_provided)
+        if sel_cls is not None:
+            if sel_cls not in plan_seq:
+                plan_seq.update(_plan(sel_cls, is_injectable, externally_provided,
+                                      True, dependency_stack))
+            type_for_arg[argname] = sel_cls
+        else:
+            if input_is_type or strict:
+                if not types:
+                    msg = "Parameter '{}' is lacking annotations in " \
+                          "'{}.__init__()'. Not possible to build a plan".format(
+                        argname, as_class_names(class_or_func))
+                else:
+                    msg = "Any of {} types are required ".format(as_class_names(types))
+                    msg += " for parameter '{}' ".format(argname)
+                    msg += " in '{}.__init__()' but none can be provided".format(
+                        as_class_names(class_or_func))
+                raise NonProvidableError(msg)
+
+
+    plan_seq[cls if input_is_type else FunctionArguments] = type_for_arg
+    return plan_seq
+
+
+def _select_type(types, is_injectable, externally_provided):
+    """ Choose the first type that can be provided. None otherwise. """
+    sel_cls = None
+    for candidate in types:
+        if is_injectable(candidate) or externally_provided(candidate):
+            sel_cls = candidate
+            break
+    return sel_cls
```

Comment:
```suggestion
    for candidate in types:
        if is_injectable(candidate) or externally_provided(candidate):
            return candidate
    return None
```
the last "return None" is not really needed as well, it is just to be more explicit.

Reply:
Much better.